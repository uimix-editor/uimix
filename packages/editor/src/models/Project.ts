/* eslint-disable @typescript-eslint/no-explicit-any */

import * as Y from "yjs";
import { posix as path } from "path-browserify";
import { getOrCreate } from "../state/Collection";
import { SelectableMap } from "./Selectable";
import { Node, NodeMap } from "./Node";
import { computed, makeObservable, ObservableMap } from "mobx";
import { ProjectJSON } from "@uimix/node-data";
import { toProjectJSON } from "./toProjectJSON";
import { ImageManager } from "./ImageManager";
import { Component } from "./Component";
import { ObservableYMap } from "../utils/ObservableYMap";

export interface PageHierarchyFolderEntry {
  type: "directory";
  path: string;
  name: string;
  children: PageHierarchyEntry[];
}

export interface PageHierarchyPageEntry {
  type: "file";
  path: string;
  name: string;
  page: Node;
}

export type PageHierarchyEntry =
  | PageHierarchyFolderEntry
  | PageHierarchyPageEntry;

class Pages {
  constructor(project: Project) {
    this.project = project;
    this.node = project.node;
    makeObservable(this);
  }

  node: Node;
  readonly project: Project;

  @computed get count(): number {
    return this.node.childCount;
  }

  get all(): Node[] {
    return this.node.children;
  }

  create(filePath: string): Node {
    const node = this.project.nodes.create("page");
    node.name = filePath;
    return node;
  }

  toHierarchy(): PageHierarchyFolderEntry {
    const root: PageHierarchyFolderEntry = {
      type: "directory",
      name: "",
      path: "",
      children: [],
    };
    const parents = new Map<string, PageHierarchyFolderEntry>();
    parents.set("", root);

    const mkdirp = (segments: string[]): PageHierarchyFolderEntry => {
      if (segments.length === 0) {
        return root;
      }

      const existing = parents.get(segments.join("/"));
      if (existing) {
        return existing;
      }

      const parent = mkdirp(segments.slice(0, -1));
      const dir: PageHierarchyFolderEntry = {
        type: "directory",
        name: segments[segments.length - 1],
        path: segments.join("/"),
        children: [],
      };
      parent.children.push(dir);
      parents.set(segments.join("/"), dir);
      return dir;
    };

    const pages = Array.from(this.all);
    pages.sort((a, b) => a.name.localeCompare(b.name));

    for (const page of pages) {
      const segments = page.name.split(path.sep);
      const parent = mkdirp(segments.slice(0, -1));

      const item: PageHierarchyPageEntry = {
        type: "file",
        name: segments[segments.length - 1],
        path: page.name,
        page,
      };
      parent.children.push(item);
    }

    return root;
  }

  affectedPagesForPath(path: string): Node[] {
    return this.all.filter(
      (page) => page.name === path || page.name.startsWith(path + "/")
    );
  }

  deletePageOrPageFolder(path: string) {
    const pagesToDelete = this.affectedPagesForPath(path);

    for (const page of pagesToDelete) {
      page.remove();
      // TODO: delete dangling nodes?
      //this.project.nodes.remove(doc.root);
    }
  }

  renamePageOrPageFolder(path: string, newPath: string) {
    if (path === newPath) {
      return;
    }

    const pagesToDelete = this.affectedPagesForPath(path);

    for (const page of pagesToDelete) {
      page.name = newPath + page.name.slice(path.length);
    }
  }
}

export class Project {
  constructor(data: Y.Map<any>) {
    this.data = ObservableYMap.get(data);
    this.nodes = new NodeMap(this);
    this.selectables = new SelectableMap(this);
    this.node = this.nodes.create("project", "project");
    this.pages = new Pages(this);
  }

  readonly data: ObservableYMap<any>;
  readonly nodes: NodeMap;
  readonly selectables: SelectableMap;
  readonly node: Node;
  readonly pages: Pages;
  readonly imageManager = new ImageManager();

  toJSON(): ProjectJSON {
    return toProjectJSON(this.node.children.map((c) => c.selectable));
  }

  loadJSON(json: ProjectJSON) {
    this.node.clear();

    for (const [id, nodeJSON] of Object.entries(json.nodes)) {
      const node = this.nodes.getOrCreate(nodeJSON.type, id);
      node.loadJSON(nodeJSON);
      if (node.type === "page") {
        this.node.append([node]);
      }
    }
    for (const [id, style] of Object.entries(json.styles)) {
      const selectable = this.selectables.get(id.split(":"));
      selectable.selfStyle.loadJSON(style);
    }
  }

  get components(): Component[] {
    const components: Component[] = [];

    for (const page of this.pages.all) {
      for (const node of page.children) {
        const component = Component.from(node);
        if (component) {
          components.push(component);
        }
      }
    }

    return components;
  }
}
