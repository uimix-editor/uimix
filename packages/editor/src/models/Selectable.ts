import { ObservableYMap } from "@uimix/foundation/src/utils/ObservableYMap";
import { abstractNodeTypes, Node } from "./Node";
import { CascadedStyle, defaultStyle, IStyle, PartialStyle } from "./Style";
import * as Y from "yjs";
import { getOrCreate } from "@uimix/foundation/src/utils/Collection";
import { computed, makeObservable, observable } from "mobx";
import { Rect } from "paintvec";
import { resizeWithBoundingBox } from "../services/Resize";
import {
  SelectableJSON,
  NodeJSON,
  NodeType,
  ProjectJSON,
} from "@uimix/node-data";
import { Project } from "./Project";
import { Component } from "./Component";

export interface IComputedRectProvider {
  readonly value: Rect | undefined;
  markDirty(): void;
}

export class StubComputedRectProvider implements IComputedRectProvider {
  constructor(rect: Rect) {
    this.value = rect;
  }
  readonly value: Rect;
  markDirty() {}
}

class SelectablePartialStyle extends PartialStyle {
  constructor(selectable: Selectable) {
    super();
    this.selectable = selectable;
  }

  selectable: Selectable;

  get data(): ObservableYMap<unknown> | undefined {
    return ObservableYMap.get(
      this.selectable.selectableMap.stylesData.get(this.selectable.id)
    );
  }

  get dataForWrite(): ObservableYMap<unknown> {
    return ObservableYMap.get(
      getOrCreate(
        this.selectable.selectableMap.stylesData,
        this.selectable.id,
        () => new Y.Map()
      )
    );
  }
}

// a node or a inner node of an instance
export class Selectable {
  constructor(project: Project, idPath: string[]) {
    if (idPath.length === 0) {
      throw new Error("idPath must not be empty");
    }
    this.project = project;
    this.selectableMap = project.selectables;
    this.idPath = idPath;
    makeObservable(this);
  }

  readonly project: Project;
  readonly selectableMap: SelectableMap;

  // Non component root nodes:
  // [outermost instance ID, ..., innermost instance ID, original node ID]
  // Component root nodes:
  // [outermost instance ID, ..., innermost instance ID]
  readonly idPath: string[];

  get id(): string {
    return this.idPath.join(":");
  }

  @computed get children(): Selectable[] {
    const mainComponent = this.mainComponent;
    if (mainComponent) {
      return mainComponent.rootNode.children.map((child) => {
        return this.project.selectables.get([...this.idPath, child.id]);
      });
    }

    return this.originalNode.children.map((child) => {
      return this.project.selectables.get([
        ...this.idPath.slice(0, -1),
        child.id,
      ]);
    });
  }

  @computed get parent(): Selectable | undefined {
    const { nodePath } = this;
    const originalNode = nodePath[nodePath.length - 1];
    const parentNode = originalNode.parent;
    if (!parentNode) {
      return;
    }

    if (nodePath.length === 1) {
      return this.project.selectables.get([parentNode.id]);
    }

    if (parentNode.isComponentRoot) {
      // looks like a component root
      return this.project.selectables.get(this.idPath.slice(0, -1));
    } else {
      return this.project.selectables.get([
        ...this.idPath.slice(0, -1),
        parentNode.id,
      ]);
    }
  }

  // children/parent ignoring abstract nodes

  @computed get offsetParent(): Selectable | undefined {
    const parent = this.parent;
    if (parent && abstractNodeTypes.includes(parent.node.type)) {
      return parent.offsetParent;
    }
    return parent;
  }

  @computed get offsetChildren(): Selectable[] {
    const children: Selectable[] = [];

    for (const child of this.children) {
      if (abstractNodeTypes.includes(child.originalNode.type)) {
        children.push(...child.children);
      } else {
        children.push(child);
      }
    }
    return children;
  }

  @computed get pageSelectable(): Selectable | undefined {
    if (this.originalNode.type === "page") {
      return this;
    }
    return this.parent?.pageSelectable;
  }

  // ancestors ([root, ..., parent, this])
  @computed get ancestors(): readonly Selectable[] {
    const result: Selectable[] = [];
    let current: Selectable | undefined = this;
    while (current) {
      result.unshift(current);
      current = current.parent;
    }
    return result;
  }

  @computed get nextSibling(): Selectable | undefined {
    const nextNode = this.originalNode.nextSibling;
    if (nextNode) {
      return this.project.selectables.get([
        ...this.idPath.slice(0, -1),
        nextNode.id,
      ]);
    }
  }

  @computed get previousSibling(): Selectable | undefined {
    const previousNode = this.originalNode.previousSibling;
    if (previousNode) {
      return this.project.selectables.get([
        ...this.idPath.slice(0, -1),
        previousNode.id,
      ]);
    }
  }

  @computed get selfStyle(): PartialStyle {
    return new SelectablePartialStyle(this);
  }

  @computed get nodePath(): Node[] {
    return this.idPath.map((id) => this.project.nodes.getOrThrow(id));
  }

  @computed get originalNode(): Node {
    const { nodePath } = this;
    return nodePath[nodePath.length - 1];
  }

  @computed get node(): Node {
    const mainComponent = this.mainComponent;
    if (mainComponent) {
      return mainComponent.rootNode;
    }
    return this.originalNode;
  }

  @computed get style(): CascadedStyle {
    return this.getStyle("displayed");
  }
  @computed get originalStyle(): CascadedStyle {
    return this.getStyle("original");
  }

  get superSelectable(): Selectable | undefined {
    if (this.idPath.length === 1) {
      const mainComponent = this.mainComponent;
      if (mainComponent) {
        return mainComponent.rootNode.selectable;
      }
      return;
    }
    return this.project.selectables.get(this.idPath.slice(1));
  }

  // resolveMainComponent=false to get main component ID of an instance
  private getStyle(type: "original" | "displayed"): CascadedStyle {
    const { nodePath } = this;

    let superStyle: IStyle;

    if (nodePath.length === 1) {
      superStyle = defaultStyle;

      if (type === "displayed") {
        const mainComponent = this.mainComponent;
        if (mainComponent) {
          superStyle = this.project.selectables
            .get([mainComponent.rootNode.id])
            .getStyle("original");
        }
      }
    } else {
      const superSelectable = this.project.selectables.get(
        this.idPath.slice(1)
      );
      superStyle = superSelectable.getStyle(type);
    }

    return new CascadedStyle(this.selfStyle, superStyle);
  }

  @computed get mainComponent(): Component | undefined {
    const originalNode = this.originalNode;
    if (originalNode.type === "instance") {
      const mainComponentID = this.originalStyle.mainComponent;
      if (mainComponentID) {
        const ownerComponents = this.nodePath.map(
          (node) => node.ownerComponent?.node
        );

        const mainComponentNode = this.project.nodes.get(mainComponentID);
        if (mainComponentNode && !ownerComponents.includes(mainComponentNode)) {
          return Component.from(mainComponentNode);
        }
      }
    }
    if (originalNode.type === "variant") {
      const componentNode = originalNode.parent;
      if (componentNode?.type === "component") {
        return Component.from(componentNode);
      }
    }
  }

  @computed private get _selected(): boolean {
    return this.selectableMap.selectionData.has(this.id);
  }

  private set _selected(value: boolean) {
    if (this._selected === value) {
      return;
    }
    if (value) {
      this.selectableMap.selectionData.set(this.id, true);
    } else {
      this.selectableMap.selectionData.delete(this.id);
    }
  }

  @computed get selected(): boolean {
    return this._selected;
  }

  select() {
    this._selected = true;
    for (const child of this.children) {
      child.deselect();
    }
  }

  deselect() {
    this._selected = false;
    for (const child of this.children) {
      child.deselect();
    }
  }

  @computed.struct get selectedDescendants(): Selectable[] {
    if (this.selected) {
      return [this];
    }
    return this.children.flatMap((child) => child.selectedDescendants);
  }

  @computed get ancestorSelected(): boolean {
    return this.selected || !!this.parent?.ancestorSelected;
  }

  @computed get ancestorLocked(): boolean {
    return this.style.locked || !!this.parent?.ancestorLocked;
  }

  @observable computedRectProvider: IComputedRectProvider | undefined =
    undefined;

  @computed.struct get computedRect(): Rect {
    return this.computedRectProvider?.value ?? new Rect();
  }

  @computed get computedPaddingRect(): Rect {
    const { computedRect, style } = this;
    const {
      borderLeftWidth,
      borderRightWidth,
      borderTopWidth,
      borderBottomWidth,
    } = style;

    return Rect.from({
      left: computedRect.left + borderLeftWidth,
      top: computedRect.top + borderTopWidth,
      width: computedRect.width - borderLeftWidth - borderRightWidth,
      height: computedRect.height - borderTopWidth - borderBottomWidth,
    });
  }

  @computed get computedContentRect(): Rect {
    const { computedRect, style } = this;
    const {
      borderLeftWidth,
      borderRightWidth,
      borderTopWidth,
      borderBottomWidth,
      paddingLeft,
      paddingRight,
      paddingTop,
      paddingBottom,
    } = style;

    return Rect.from({
      left: computedRect.left + borderLeftWidth + paddingLeft,
      top: computedRect.top + borderTopWidth + paddingTop,
      width:
        computedRect.width -
        borderLeftWidth -
        borderRightWidth -
        paddingLeft -
        paddingRight,
      height:
        computedRect.height -
        borderTopWidth -
        borderBottomWidth -
        paddingTop -
        paddingBottom,
    });
  }

  @computed get computedMarginRect(): Rect {
    if (this.isAbsolute) {
      return this.computedRect;
    }

    const { computedRect, style } = this;
    const { marginLeft, marginRight, marginTop, marginBottom } = style;

    return Rect.from({
      left: computedRect.left - marginLeft,
      top: computedRect.top - marginTop,
      width: computedRect.width + marginLeft + marginRight,
      height: computedRect.height + marginTop + marginBottom,
    });
  }

  @computed get computedOffsetRect(): Rect {
    const { offsetParent } = this;
    if (!offsetParent) {
      return this.computedRect;
    }
    return this.computedRect.translate(
      offsetParent.computedPaddingRect.topLeft.neg
    );
  }

  @computed get computedOffsetLeft(): number {
    const { offsetParent } = this;
    if (offsetParent) {
      return this.computedRect.left - offsetParent.computedPaddingRect.left;
    }
    return this.computedRect.left;
  }

  @computed get computedOffsetTop(): number {
    const { offsetParent } = this;
    if (offsetParent) {
      return this.computedRect.top - offsetParent.computedPaddingRect.top;
    }
    return this.computedRect.top;
  }

  @computed get computedOffsetRight(): number | undefined {
    const { offsetParent } = this;
    if (offsetParent) {
      return offsetParent.computedPaddingRect.right - this.computedRect.right;
    }
  }

  @computed get computedOffsetBottom(): number | undefined {
    const { offsetParent } = this;
    if (offsetParent) {
      return offsetParent.computedPaddingRect.bottom - this.computedRect.bottom;
    }
  }

  @observable collapsed = false;

  expandAncestors() {
    const { parent } = this;
    if (parent) {
      parent.expandAncestors();
      parent.collapsed = false;
    }
  }

  @computed get inFlowChildren(): Selectable[] {
    return this.children.filter((child) => !child.isAbsolute);
  }

  @computed get isAbsolute(): boolean {
    if (this.parent?.style.layout !== "none") {
      return this.style.absolute;
    }
    return true;
  }

  createBefore(type: NodeType, next: Selectable | undefined): Selectable {
    const node = this.project.nodes.create(type);
    this.originalNode.insertBefore([node], next?.originalNode);
    return this.project.selectables.get([node.id]);
  }

  append(type: NodeType): Selectable {
    return this.createBefore(type, undefined);
  }

  // insert(index: number, contents: Omit<NodeJSON, "id">[]): Selectable[] {
  //   this.node.insert(index, contents);
  //   return this.children.slice(index, index + contents.length);
  // }

  // prepend(contents: Omit<NodeJSON, "id">[]): Selectable[] {
  //   return this.insert(0, contents);
  // }

  // append(contents: Omit<NodeJSON, "id">[]): Selectable[] {
  //   return this.insert(this.children.length, contents);
  // }

  includes(other: Selectable): boolean {
    return other.ancestors.includes(this);
  }

  get canInsertChild(): boolean {
    const { originalNode } = this;
    return originalNode.type === "page" || originalNode.type === "frame";
  }

  @computed.struct get usedFontFamilies(): Set<string> {
    const { style } = this;
    const result = new Set<string>();
    if (style.fontFamily) {
      result.add(style.fontFamily);
    }

    for (const child of this.children) {
      for (const font of child.usedFontFamilies) {
        result.add(font);
      }
    }
    return result;
  }

  @computed get refID(): string | undefined {
    const ownerComponent = this.originalNode.ownerComponent;
    if (!ownerComponent || ownerComponent.rootNode === this.originalNode) {
      return undefined;
    }

    return ownerComponent.refIDs.get(this.originalNode.id);
  }

  /// mutations

  remove() {
    if (this.idPath.length >= 2) {
      // TODO: hide?
      throw new Error(
        "TODO: removing items inside an instance is not supported yet"
      );
    }

    const { originalNode } = this;
    const { parent } = originalNode;
    if (originalNode.type !== "variant" && parent?.type === "component") {
      // removing component root node -> remove component
      parent.remove();
      return;
    }

    originalNode.remove();
  }

  insertBefore(
    selectables: readonly Selectable[],
    dstNextSibling: Selectable | undefined,
    {
      fixPosition = true,
    }: {
      fixPosition?: boolean;
    } = {}
  ) {
    selectables = selectables.filter((s) => !s.includes(this));
    if (selectables.length === 0) {
      return;
    }

    if (
      this.idPath.length > 1 ||
      selectables.some((selectable) => selectable.idPath.length > 1)
    ) {
      console.log(
        'TODO: moving items inside an instance is not supported yet"'
      );
      return;
    }

    this.node.insertBefore(
      selectables.map((s) => s.originalNode),
      dstNextSibling?.originalNode
    );

    if (fixPosition) {
      for (const selectable of selectables) {
        const absolute =
          this.style.layout === "none" || selectable.style.absolute;

        if (absolute) {
          const bbox = selectable.computedRect;
          resizeWithBoundingBox(selectable, bbox, { x: true, y: true });
        }
      }
    }
  }

  toJSON(): SelectableJSON {
    const originalNode = this.originalNode;
    const node = this.node;

    return {
      id: this.id,
      name: originalNode.name,
      type: node.type,
      original: {
        id: originalNode.id,
        type: originalNode.type,
        condition: originalNode.condition,
      },
      style: this.style.toJSON(),
      selfStyle: this.selfStyle.toJSON(),
      children: this.children.map((child) => child.toJSON()),
    };
  }

  static fromJSON(project: Project, json: SelectableJSON): Selectable {
    const node = project.nodes.create(json.type);
    node.name = json.name;
    const selectable = node.selectable;
    selectable.selfStyle.loadJSON(json.style);

    const children = json.children.map((child) =>
      Selectable.fromJSON(project, child)
    );
    node.insertBefore(
      children.map((c) => c.originalNode),
      undefined
    );
    return selectable;
  }
}

export class SelectableMap {
  constructor(project: Project) {
    this.project = project;
  }

  get stylesData(): ObservableYMap<Y.Map<unknown>> {
    return ObservableYMap.get(this.project.doc.getMap("styles"));
  }
  get selectionData(): ObservableYMap<true> {
    return ObservableYMap.get(this.project.doc.getMap("selection"));
  }

  private readonly project: Project;
  private readonly selectableMap = new Map<string, Selectable>();

  get(idPath: string[]): Selectable {
    return getOrCreate(this.selectableMap, idPath.join(":"), () => {
      return new Selectable(this.project, idPath);
    });
  }
}

// TODO generate correctly from instance contents
export function selectablesToProjectJSON(
  selectables: Selectable[]
): ProjectJSON {
  const nodeJSONs: Record<string, NodeJSON> = {};
  const styles: Record<string, Partial<IStyle>> = {};

  const addRecursively = (selectable: Selectable) => {
    if (selectable.idPath.length === 1) {
      const node = selectable.originalNode;
      nodeJSONs[node.id] = node.toJSON();
      if (selectables.includes(selectable)) {
        nodeJSONs[node.id].parent = undefined;
      }
    }

    styles[selectable.id] = selectable.selfStyle.toJSON();
    for (const child of selectable.children) {
      addRecursively(child);
    }
  };

  for (const selected of selectables) {
    addRecursively(selected);
  }

  return {
    nodes: nodeJSONs,
    styles,
  };
}
