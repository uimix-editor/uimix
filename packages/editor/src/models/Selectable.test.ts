import { describe, expect, it } from "vitest";
import * as Y from "yjs";
import { Project } from "./Project";
import { selectablesToProjectJSON } from "./Selectable";
import * as path from "path";
import { ProjectFiles } from "../../../cli/src/compiler/ProjectFiles";

describe(selectablesToProjectJSON.name, () => {
  it("works", () => {
    const rootDir = path.resolve("../sandbox");
    const projectFiles = ProjectFiles.load(rootDir, {
      filePattern: "src/components.uimix",
    });
    const projectJSON = projectFiles.projectJSON;

    const ydoc = new Y.Doc();
    const project = new Project(ydoc);
    project.loadJSON(projectJSON);

    const partial = selectablesToProjectJSON(
      project.pages.all[0].selectable.children
    );

    expect(partial).toMatchSnapshot();
  });
});
