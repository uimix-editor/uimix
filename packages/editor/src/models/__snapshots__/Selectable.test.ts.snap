// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`selectablesToProjectJSON > works 1`] = `
{
  "nodes": {
    "1Bvi2z3TfkSpYXW29daTN8": {
      "condition": undefined,
      "index": 0,
      "name": "Stack",
      "parent": "1Bvi2zJALza4DrqmNaGYAk",
      "type": "frame",
    },
    "1Bvi2z3To8ddXo6R6ZftVD": {
      "condition": undefined,
      "index": 1,
      "name": "Item 1",
      "parent": "1Bvi2z3TfkSpYXW29daTN8",
      "type": "frame",
    },
    "1Bvi2z3TpKXtTJLTRPTEd8": {
      "condition": undefined,
      "index": 0,
      "name": "Item 0",
      "parent": "1Bvi2z3TfkSpYXW29daTN8",
      "type": "frame",
    },
    "1Bvi2z3U3SiK6r7gf9vh7N": {
      "condition": undefined,
      "index": 3,
      "name": "Text",
      "parent": "1Bvi2z3TfkSpYXW29daTN8",
      "type": "text",
    },
    "1Bvi2z3UB1xUVg4Xz9WTpJ": {
      "condition": undefined,
      "index": 2,
      "name": "Item 2",
      "parent": "1Bvi2z3TfkSpYXW29daTN8",
      "type": "frame",
    },
    "1Bvi2z3UGTM64JtRT7Qc1S": {
      "condition": undefined,
      "index": 0,
      "name": "Text",
      "parent": "1Bvi2z3UPfi6W3L8nDt5VA",
      "type": "text",
    },
    "1Bvi2z3UPfi6W3L8nDt5VA": {
      "condition": undefined,
      "index": 0,
      "name": "",
      "parent": "1Bvi2z3USBZi1pWrvPd6Ut",
      "type": "frame",
    },
    "1Bvi2z3USBZi1pWrvPd6Ut": {
      "condition": undefined,
      "index": 11,
      "name": "Button",
      "parent": undefined,
      "type": "component",
    },
    "1Bvi2z3UVsEskiSt2TcdL7": {
      "condition": {
        "type": "hover",
      },
      "index": 1,
      "name": "",
      "parent": "1Bvi2z3USBZi1pWrvPd6Ut",
      "type": "variant",
    },
    "1Bvi2z3UaHFmaikKtfcmGX": {
      "condition": {
        "type": "maxWidth",
        "value": 767,
      },
      "index": 2,
      "name": "",
      "parent": "1Bvi2z3USBZi1pWrvPd6Ut",
      "type": "variant",
    },
    "1Bvi2z3UrU9VkawwAteXs8": {
      "condition": undefined,
      "index": 1,
      "name": "Instance",
      "parent": "1Bvi365mtWZmXKDPeJTB4s",
      "type": "instance",
    },
    "1Bvi2zJALza4DrqmNaGYAk": {
      "condition": undefined,
      "index": 13,
      "name": "Stack",
      "parent": undefined,
      "type": "component",
    },
    "1Bvi319YVxZELz4tEskN1S": {
      "condition": undefined,
      "index": 0,
      "name": "Frame",
      "parent": "1Bvi3BSSb2oaHb1ZqUEk6H",
      "type": "frame",
    },
    "1Bvi31VKu8gfy6XzJ8xrHA": {
      "condition": undefined,
      "index": 1,
      "name": "Header",
      "parent": "1Bvi319YVxZELz4tEskN1S",
      "type": "foreign",
    },
    "1Bvi32d9RMi3ES5kk2UvuM": {
      "condition": undefined,
      "index": 0,
      "name": "Button",
      "parent": "1Bvi365mtWZmXKDPeJTB4s",
      "type": "foreign",
    },
    "1Bvi365mtWZmXKDPeJTB4s": {
      "condition": undefined,
      "index": 3,
      "name": "Frame",
      "parent": "1Bvi319YVxZELz4tEskN1S",
      "type": "frame",
    },
    "1Bvi392qBdMy93LGLcVymb": {
      "condition": undefined,
      "index": 2,
      "name": "Image",
      "parent": "1Bvi365mtWZmXKDPeJTB4s",
      "type": "image",
    },
    "1Bvi39mct9UNYZ9eTCCRdV": {
      "condition": undefined,
      "index": 3,
      "name": "Stack",
      "parent": "1Bvi365mtWZmXKDPeJTB4s",
      "type": "instance",
    },
    "1Bvi3BSSb2oaHb1ZqUEk6H": {
      "condition": undefined,
      "index": 14,
      "name": "Page",
      "parent": undefined,
      "type": "component",
    },
  },
  "styles": {
    "1Bvi2z3TfkSpYXW29daTN8": {
      "fills": [
        {
          "hex": "#FFFFFF",
          "type": "solid",
        },
      ],
      "gap": 10,
      "height": {
        "type": "hugContents",
      },
      "layout": "stack",
      "paddingBottom": 30,
      "paddingLeft": 40,
      "paddingRight": 20,
      "paddingTop": 10,
      "position": {
        "x": {
          "start": 50,
          "type": "start",
        },
        "y": {
          "start": 184,
          "type": "start",
        },
      },
      "width": {
        "type": "hugContents",
      },
    },
    "1Bvi2z3To8ddXo6R6ZftVD": {
      "fills": [
        {
          "hex": "#22C55E",
          "type": "solid",
        },
      ],
      "height": {
        "type": "fixed",
        "value": 80,
      },
      "width": {
        "type": "fixed",
        "value": 40,
      },
    },
    "1Bvi2z3TpKXtTJLTRPTEd8": {
      "fills": [
        {
          "hex": "#EF4444",
          "type": "solid",
        },
      ],
      "height": {
        "type": "fixed",
        "value": 50,
      },
      "width": {
        "type": "fixed",
        "value": 50,
      },
    },
    "1Bvi2z3U3SiK6r7gf9vh7N": {
      "fills": [
        {
          "hex": "#000000",
          "type": "solid",
        },
      ],
      "fontSize": 20,
      "height": {
        "type": "hugContents",
      },
      "textContent": "Hello, world!",
      "width": {
        "type": "hugContents",
      },
    },
    "1Bvi2z3UB1xUVg4Xz9WTpJ": {
      "fills": [
        {
          "hex": "#3B82F6",
          "type": "solid",
        },
      ],
      "height": {
        "type": "fixed",
        "value": 40,
      },
      "width": {
        "type": "fixed",
        "value": 80,
      },
    },
    "1Bvi2z3UGTM64JtRT7Qc1S": {
      "fills": [
        {
          "hex": "#111827",
          "type": "solid",
        },
      ],
      "height": {
        "type": "hugContents",
      },
      "textContent": "Button",
      "width": {
        "type": "hugContents",
      },
    },
    "1Bvi2z3UPfi6W3L8nDt5VA": {
      "fills": [
        {
          "hex": "#E5E7EB",
          "type": "solid",
        },
      ],
      "height": {
        "type": "hugContents",
      },
      "layout": "stack",
      "paddingBottom": 4,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingTop": 4,
      "position": {
        "x": {
          "start": 50,
          "type": "start",
        },
        "y": {
          "start": 400,
          "type": "start",
        },
      },
      "width": {
        "type": "hugContents",
      },
    },
    "1Bvi2z3USBZi1pWrvPd6Ut": {},
    "1Bvi2z3UVsEskiSt2TcdL7": {
      "fills": [
        {
          "hex": "#3B82F6",
          "type": "solid",
        },
      ],
      "position": {
        "x": {
          "start": 200,
          "type": "start",
        },
        "y": {
          "start": 400,
          "type": "start",
        },
      },
    },
    "1Bvi2z3UVsEskiSt2TcdL7:1Bvi2z3UGTM64JtRT7Qc1S": {
      "fills": [
        {
          "hex": "#FFFFFF",
          "type": "solid",
        },
      ],
    },
    "1Bvi2z3UaHFmaikKtfcmGX": {
      "position": {
        "x": {
          "start": 350,
          "type": "start",
        },
        "y": {
          "start": 400,
          "type": "start",
        },
      },
    },
    "1Bvi2z3UaHFmaikKtfcmGX:1Bvi2z3UGTM64JtRT7Qc1S": {},
    "1Bvi2z3UrU9VkawwAteXs8": {
      "fills": [
        {
          "hex": "#BABABA",
          "type": "solid",
        },
      ],
      "mainComponent": "1Bvi2z3USBZi1pWrvPd6Ut",
      "position": {
        "x": {
          "start": 160,
          "type": "start",
        },
        "y": {
          "start": 28,
          "type": "start",
        },
      },
    },
    "1Bvi2z3UrU9VkawwAteXs8:1Bvi2z3UGTM64JtRT7Qc1S": {
      "textContent": "Override Name",
    },
    "1Bvi2zJALza4DrqmNaGYAk": {},
    "1Bvi319YVxZELz4tEskN1S": {
      "fills": [
        {
          "hex": "#FFFFFF",
          "type": "solid",
        },
      ],
      "height": {
        "type": "fixed",
        "value": 456,
      },
      "layout": "stack",
      "position": {
        "x": {
          "start": 50,
          "type": "start",
        },
        "y": {
          "start": 585,
          "type": "start",
        },
      },
      "stackAlign": "center",
      "stackDirection": "y",
      "width": {
        "type": "fixed",
        "value": 744,
      },
    },
    "1Bvi31VKu8gfy6XzJ8xrHA": {
      "foreignComponent": {
        "name": "Header",
        "path": "src/stories/Header.tsx",
        "props": {},
        "type": "react",
      },
      "height": {
        "type": "hugContents",
      },
      "width": {
        "type": "fillContainer",
        "value": 744,
      },
    },
    "1Bvi32d9RMi3ES5kk2UvuM": {
      "foreignComponent": {
        "name": "Button",
        "path": "src/stories/Button.tsx",
        "props": {
          "backgroundColor": "red",
          "label": "Button1",
          "primary": true,
          "size": "large",
        },
        "type": "react",
      },
      "height": {
        "type": "hugContents",
      },
      "position": {
        "x": {
          "start": 33,
          "type": "start",
        },
        "y": {
          "start": 28,
          "type": "start",
        },
      },
      "width": {
        "type": "hugContents",
      },
    },
    "1Bvi365mtWZmXKDPeJTB4s": {
      "fills": [
        {
          "hex": "#FFFFFF",
          "type": "solid",
        },
      ],
      "height": {
        "type": "fillContainer",
        "value": 173,
      },
      "position": {
        "x": {
          "start": 210,
          "type": "start",
        },
        "y": {
          "start": 142,
          "type": "start",
        },
      },
      "width": {
        "type": "fillContainer",
        "value": 428,
      },
    },
    "1Bvi392qBdMy93LGLcVymb": {
      "fills": [
        {
          "hex": "#FFFFFF",
          "type": "solid",
        },
      ],
      "height": {
        "type": "fixed",
        "value": 100,
      },
      "imageHash": "YmY-tuteU03QY-k_2ARy1uacNmXDNmOX-FFAlj2qCIo=",
      "position": {
        "x": {
          "start": 33,
          "type": "start",
        },
        "y": {
          "start": 97,
          "type": "start",
        },
      },
      "width": {
        "type": "fixed",
        "value": 100,
      },
    },
    "1Bvi39mct9UNYZ9eTCCRdV": {
      "mainComponent": "1Bvi2zJALza4DrqmNaGYAk",
      "position": {
        "x": {
          "start": 165,
          "type": "start",
        },
        "y": {
          "start": 86,
          "type": "start",
        },
      },
    },
    "1Bvi39mct9UNYZ9eTCCRdV:1Bvi2z3To8ddXo6R6ZftVD": {},
    "1Bvi39mct9UNYZ9eTCCRdV:1Bvi2z3TpKXtTJLTRPTEd8": {},
    "1Bvi39mct9UNYZ9eTCCRdV:1Bvi2z3U3SiK6r7gf9vh7N": {},
    "1Bvi39mct9UNYZ9eTCCRdV:1Bvi2z3UB1xUVg4Xz9WTpJ": {},
    "1Bvi3BSSb2oaHb1ZqUEk6H": {},
  },
}
`;
